<a href="https://colab.research.google.com/github/miriamm172/TECNICAS-DE-CLASIFICACION/blob/master/Random_Forest_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**BOSQUES ALEATORIOS**: Ejemplo de clasificación del dataset [Pima Indian Diabetes](https://https://colab.research.google.com/drive/1NddH8HNl4pt665rqQFF0PU81y39avDKT?authuser=1#scrollTo=IBSKRPYjQj6R&line=1&uniqifier=1) para predecir si un paciente tiene diabetes o no. El datasert consta de 9 columnas y 768 muestras.

Importamos las librerías necesarias y cargamos el documento que contiene las muestras.

In [ ]:
import pandas as pd
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split 

In [ ]:
col_names = ['pregnant', 'glucose', 'bp', 'skin', 'insulin', 'bmi', 'pedigree', 'age', 'label'] #designamos nombres de las columnas
dataset = pd.read_csv("pima-indians-diabetes.csv", header=None, names=col_names) #colocamos su nombre a cada columa "names"
dataset = dataset.iloc [1:]
dataset.head()

pregnant glucose  bp skin insulin   bmi pedigree age label
1        6     148  72   35       0  33.6    0.627  50     1
2        1      85  66   29       0  26.6    0.351  31     0
3        8     183  64    0       0  23.3    0.672  32     1
4        1      89  66   23      94  28.1    0.167  21     0
5        0     137  40   35     168  43.1    2.288  33     1

Dividimos la data: "X" variables independietes (características) y "y" variables dependientes (etiqueta).
Separamos los datos, 70% para entrenamiento y 30% para pueba.

In [ ]:
feature_cols = ['pregnant', 'insulin', 'bmi', 'age','glucose','bp','pedigree']
X = dataset[feature_cols] # características
y = dataset.label # etiquetas de salida

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=1)

Definimos del modelo de bosques aleatorios de clasificación, lo entrenamos (datos de entrenamiento) y generamos las nuevas predicciones (datos prueba).

In [ ]:
clf = RandomForestClassifier(n_estimators=100, random_state=0) #modelo, "n_estimators" define el numero de árboles en el bosque
clf = clf.fit(X_train,y_train) #entrenamiento
y_pred = clf.predict(X_test) #nuevas predicciones

Midiendo la eficacia del modelo a través de la matriz de confusión.

In [ ]:
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score
print("Matriz de confusión")
print(confusion_matrix(y_test,y_pred))
print("Métricas de clasificación")
print(classification_report(y_test,y_pred))

Matriz de confusión
[[131  15]
 [ 30  55]]
Métricas de clasificación
              precision    recall  f1-score   support

           0       0.81      0.90      0.85       146
           1       0.79      0.65      0.71        85

    accuracy                           0.81       231
   macro avg       0.80      0.77      0.78       231
weighted avg       0.80      0.81      0.80       231




*   True Positives 131
*   True Negatives 55
*   False Positives 30
*   False Negatives 15

Precisión del modelo: 81%